include/roaring/roaring_array.h

```diff
@@ -72,6 +72,19 @@ void ra_clear_without_containers(roaring_array_t *r);
  */
 void ra_clear_containers(roaring_array_t *ra);
 
+/**
+ * Make it so that a crash is likely if a container is used after an operation
+ * partially ran, generating undefined results.
+ *
+ * (API users should only encounter such states when using core versions
+ * that return error conditions vs. panic.  So it would be redundant to add
+ * checks for a "corrupt flag" on each API call that would panic if set.)
+ */
+inline static void ra_mark_corrupt(roaring_array_t *ra) {
+    assert(ra->containers == NULL);  // all containers must be free
+    ra->containers = (container_t**)((intptr_t)0xDECAFBAD);
```

Comment:
This seems odd. Is this something we find in other major important code bases?

Reply:
Depending on what you want to call precedent... there's the "blue screen of death".  Anything generally in the notion that it's better to trigger things to stop or crash than to continue on with invalid data.

I mentioned [the three possibilities here](https://github.com/RoaringBitmap/CRoaring/issues/174#issuecomment-721523782).

As it happens, I've had to add a flag for this state so the C++ destructor can run the clear() function.  But the flag isn't paid for on checking at every entry point.  One could do so and then error, but that seems an unnecessary expense for this edge case.

Reply:
I am not arguing the purpose, I am asking about how you reassign it from NULL to 0xDECAFBAD. I have never seen this strategy before and I am not sure I understand the reasoning behind it.

Reply:
If it were left as NULL it would be a valid empty container, and this is to make it invalid.

If you are not familiar with 0xDECAFBAD, it's just a traditionally used "easy to see in hex dumps" garbage value.

https://nedbatchelder.com/text/hexwords.html

Reply:
@AE1020  I know what hex words are... but I am still puzzled. 

Can you give a user-facing example where this would protect the user? Presumably, you have in mind a case where if we did not do this assignment, then something bad would happen.

I am not seeing the scenario in question.

In effect, I would think that we don't want to do something in the hope that it might be protective. We want to do something that we can test out and see that it does protect us.

Reply:
The bad thing is believing that your operation (in this case, an XOR) produced an empty result, when there was a memory allocation failure.

    if (!roaring_bitmap_xor_inplace_completed(x1,x2)) {
        // memory must be low, ask the GC to reclaim space
        garbage_collect();
        // that may take a while, run interrupts to check for cancellation
        run_interrupts();
        ....
    }

The question is if there is some amount of forgetfulness in the fact that x1 doesn't represent anything meaningful.  If you forget, I say it's better to raise the odds of crashing than continuing, if there are cheap ways of doing so.

Now, what is actually checked for meaningfulness is the ROARING_FLAG_INDETERMINATE:

https://github.com/RoaringBitmap/CRoaring/pull/267/commits/28ec27cde685ffff35bebfb6890b278b42705e29#diff-0e67f5a62cae4679dbfed795bee9a7f80a7195f23cf7c1b7b4f340064e08e581R424

If this edge case were addressed "rigorously" then you would at each API entry point make sure you weren't dealing with ROARING_FLAG_INDETERMINATE (which was added later).  I deemed that overkill for this--but I don't think it's a good idea to treat it as an empty set like nothing has happened.

You either treat it as an empty set, pay for flag checks at a lot of entry points, or use the inexpensive but effective option of throwing it a curveball.  Out of the three, that would be my choice.  You might be able to do it other ways that force null dereference (e.g. giving it a large size, leaving the pointer NULL).  The "idiom" of a trash hex pointer is just simple and familiar.

Reply:
So in this example...

```C++
if (!roaring_bitmap_xor_inplace_completed(x1,x2)) {
    // memory must be low, ask the GC to reclaim space
    garbage_collect();
    // that may take a while, run interrupts to check for cancellation
    run_interrupts();
    ....
}
```

The function has made x1 invalid. You want that, with some probability, access to `x1` will crash. I don't think that assigning a crazy address ensures that this address plus some offset will crash. You are just pretty sure that it will crash because the memory space is vast, and you are unlikely to be unlucky enough that your fixed address plus some offset will end up overlapping a valid memory address.

Why not just take a pointer to x1 and reassign x1 to NULL. Then the reference is gone for good and you are done with it...

```C++
if (!roaring_bitmap_xor_inplace_completed(&x1,x2)) {
  // here x1 == NULL
    ....
}
```

Reply:
> Why not just take a pointer to x1 and reassign x1 to NULL.

That might be sensible if you were guaranteed that the pointer you're being passed was one you could free and clobber the client's copy.

...but...the model you're working with is that you have a non-opaque struct that can be resident in client-owned memory, that may or may not be dynamically allocated (e.g. the C++ object doesn't do an allocation)

And if someone uses their own allocator...and you nulled out their pointer in this process...they'd have no way to free it.

> You want that, with some probability, access to x1 will crash. I don't think that assigning a crazy address ensures that this address plus some offset will crash

While it doesn't ensure it...it's a defensive move that covers most all cases.  Any formal solution for how to handle an invalid block of client-owned memory will have an associated cost; which (again) I do not think this case warrants.  I just consider it better than doing nothing.

Reply:
I realized that the name is somewhat triggering, as its from a historical path where the need to be able to tell whether the structure was in this state had not yet been absorbed.

[This renames it from ra_mark_corrupt() to ra_mark_indeterminate()](https://github.com/RoaringBitmap/CRoaring/pull/267/commits/6b0a370ca5af2ba90dc7df618880b8686939e0f9), and changes some of the points of emphasis.

Important to remember in the malloc()'-ing and free()'-ing world of C, the aren't guarantees when you break contracts.  Dereferencing NULL also doesn't *guarantee* a crash:

[https://stackoverflow.com/questions/6793262/why-dereferencing-a-null-pointer-is-undefined-behaviour](https://stackoverflow.com/questions/6793262/why-dereferencing-a-null-pointer-is-undefined-behaviour)

So even if you were to start out the process by allocating a global "bigalloc" for a single container, and then set the container inside that to NULL, and use that for every indeterminate struct... it's just additional work that obscures what's going on.

Reply:
Seems sensible.